### ISSUE TITLE
Customize filename while using rolling policies

### ISSUE THREAD

**Issue by Deilan** [serilog/serilog-sinks-file#243]
Requesting a capability to customize a filename while using [rolling policies](https://github.com/serilog/serilog-sinks-file#rolling-policies).

**Comment by zivfuture** [comment]
any update?

**Comment by osnipezzini** [comment]
Any News?

**Comment by bartelink** [comment]
An 'update' and/or 'news' would require a) an agreed spec b) a PR implementing that

- The first bit can be helped by people providing details of their requirements and/or a proposed API - this is a reasonable place for such a discussion to take place (given this sink has very broad usage, so it's worth getting the design right such that it covers the needs of as many people as possible; removing misfeatures and/or changing behavior afterward is not going to be viable).
- The second bit is a matter of someone somewhere making the time to implement it.

**Comment by ReikoTigon** [comment]
So it's not possible to let the rolling policy save the file in another folder than the original one? Because I really would need that to keep it organized.

**Comment by cocowalla** [comment]
@ReikoTigon maybe this plugin would work for you: [cocowalla/serilog-sinks-file-archive](https://github.com/cocowalla/serilog-sinks-file-archive) ?

**Comment by ReikoTigon** [comment]
Hmm if i get the plugin correctly it moves them instead of deleting them?
What I want is use the defualt mechanism by srilog to apply the daily rolling pollicy with 7 days, but tho move those old 7 logs to a dedicated "logs" folder, whereas the current log is located in the main folder.

or be able to name the current log sth like "current.log" and all the old logs "log_DATE.log" like it would do atm.

**Comment by wzuqui** [comment]
in NLog there are some properties like achiveEvery and achiveDateFormat that only modifies archived files, can it be possible to customize the same for serilog-skins-file?

see: https://nlog-project.org/documentation/v3.2.1/html/Properties_T_NLog_Targets_FileTarget.htm

**Comment by wzuqui** [comment]
In NLog, there are some properties such as "archiveEvery" and "archiveDateFormat" that only modify archived files. Is it possible to customize the same for serilog-sinks-file?

see: https://nlog-project.org/documentation/v3.2.1/html/Properties_T_NLog_Targets_FileTarget.htm

**Comment by nblumhardt** [comment]
Hi folks!

Most of this sink's heavy lifting (formatting and so on) is implemented by Serilog (`MessageTemplateTextFormatter`), or by plugged in formatters like those from _Serilog.Expressions_ (`ExpressionTemplate`) or _Serilog.Formatting.Compact_ (`CompactJsonFormatter`).

It's fairly straightforward to wire these things up in a new sink that behaves in one of the ways suggested so far in this thread. Sinks are just simple classes that implement `ILogEventSink`, which is itself a simple interface with a method `Emit(LogEvent)`.

The hesitation to add more filename rolling styles and path formatting here is because although it's not a lot of code to construct a sink for one of these cases targeting one platform, the complexity of trying to support everything in the one sink is likely to turn it into a much heavier package than what we have today, and there's not a lot of enthusiasm for that among the Serilog maintainers as far as I can tell.

If anyone's keen to implement their own specific case as a separate sink class, using formatters and so on from the packages mentioned above, and runs into trouble I'd be very happy to help - **dropping a question on Stack Overflow tagged `serilog` will get my eyes and many others on it**.

Unfortunately although it's nice to have a place to comment, this ticket is likely to be unproductive and noisy so I'll limit further activity to updates from maintainers. If anyone out there makes progress on either an implementation of a standalone sink with different rolling or formatting behavior, or a fork that adds more sophisticated configuration, we can post some links to their work here.

**Comment by TooManyLeaves** [comment]
If customizing filenames is a requirement, and you have the option of choosing your logging framework, you might consider taking a look at [NLog](https://github.com/nlog/NLog/wiki/File-target), instead.

**Comment by bartelink** [comment]
There is also a fork of this Sink, [`Serilog.Sinks.PersistentFile`](https://github.com/dfacto-lab/serilog-sinks-file) that was started with the explicit aim of implementing some of these facilities; see https://github.com/serilog/serilog-sinks-file/issues/40 for the discussion that spawned it

**Comment by yoav-melamed** [comment]
Hi,
I wrote an 'AsyncFileSink' that you can control the rolling file name. 
In addition, the sink includes the following features:
- Embedded asynchronous log writing using the `Channel<T>` API
- Control over the channel capacity (default is 50,000)
- Advanced rolling policies
   - Rolling by file size
   - Rolling on startup
   - Roll to archive folder
   - Delete old rolled files by age

You can log an issue in my repo for more features and requests

Would be happy to get your feedback:
https://github.com/yoav-melamed/serilog-sinks-asyncfile/

**Comment by pampua84** [comment]
Hi,
if I understand correctly, it is currently not possible to customize the name of the rolling file using a property in the log configuration or an environment variable, such as including the machine name:

> log-${MACHINENAME}-.txt

Is there a possibility that this functionality will be added in the future? Can an implementation proposal be made?
At the moment, it is only possible to achieve this programmatically, correct?

Thanks, and sorry for the many questions.

**Comment by bartelink** [comment]
Yes, environment strings are not expanded. Note there are multiple aspects here
- this ticket is mainly about integrating the rolling policies with custom naming, i.e. if its a log per day how does that affect naming
- the other potential need is arbitrary custom naming outside of rolling policies
- finally, the base name that's passed in can be manipulated outside the config piece - i.e. in your app, you may eb able to supply a base filename that's derived from e.g. expanding environment strings

The main point here is that any proposal needs to solve across all the needs - the impl will be the easy bit after reconciling how the rolling policy vs other factors are reconciled.

The biggest elephant in the room is that the download numbers tell you this is an extremely widespread usage sink, so adding lots of fiddly config options is off the table - you're adding noise to basic config affecting a very large number of developers if a lot of options suddenly need to be considered (that does not make it undoable, it just makes it not a five minute job to formulate a good proposal)

**Comment by pampua84** [comment]
@nblumhardt Yes I solved it this way, reading some ad hoc settings on the appsettings and not the standard ones of Serilog. Thanks

**Comment by osnipezzini** [comment]
> [@pampua84](https://github.com/pampua84) FWIW, switching to configuration in C# makes these kinds of things much easier:
> 
> .WriteTo.File(
>         $"log-{Environment.GetEnvironmentVariable("MACHINENAME")}-.txt",
>         rollingInterval: RollingInterval.Day)

Mine is configured as stated, but it still doesn't respect the chosen name.

In other words, I don't want the date to be included in the name of the current file, only in the backups.
![Image](https://github.com/user-attachments/assets/170ed904-a31d-4b3e-a8a9-ba31b74ffdd4)
